src/microbe_stage/components/OrganelleContainer.cs

```diff
@@ -640,7 +640,11 @@ public static void FetchLayoutOrganelleComponents(this ref OrganelleContainer co
             {
                 if (organelleComponent is SlimeJetComponent slimeJetComponent)
                 {
-                    if (!slimeJetComponent.IsMucocyst)
+                    if (slimeJetComponent.IsMucocyst)
+                    {
+                        ++container.MucocystCount;
```

Comment:
Isn't this updated from somewhere else? Because if this code is now newly added wouldn't that mean that mucocysts couldn't have functioned before, but they did (`CalculateOrganelleLayoutStatistics`)?

Reply:
There are 2 functions: `CalculateOrganelleLayoutStatistics` and  `FetchLayoutOrganelleComponents`. It was updated in the first one but someone forgot to add it in the second one, so many microbes didn't know they had mucocyst until their organelles split. We were able to use the mucocyst as the player because the first function was called upon leaving the editor

Reply:
I'm almost certain that cannot be the case, because that would mean that a bunch of other stuff like toxin vacuoles would be bugged similarly. `MicrobeReproductionSystem` calls `OnOrganellesChanged` which in turn calls `CalculateOrganelleLayoutStatistics`.

Reply:
Yes, sorry, of course I meant `MucocystCount` instead of `isMucocyst`. And you're right about counting it twice, i need to add `container.MucocystCount = 0;` at the beggining of the `FetchLayoutOrganelleComponents`

Reply:
I did some testing and I'm nearly certain that this change is unnecessary. I put breakpoints and checked that `CalculateOrganelleLayoutStatistics` already sets the mucocyst count, and when the code for running `FetchLayoutOrganelleComponents` is triggered I see this:
![Kuvakaappaus - 2024-11-29 12-43-52](https://github.com/user-attachments/assets/d5128246-81dc-49f9-9182-9f07368f524e)

So the count is already set and your changed code just calculates it again, which is unnecessary.

So either the current code without these changes is correct or you've found a rare bug I was unable to trigger. So please re-check this. I've now checked the rest of the PR and I think it's fine for merge after the changes to OrganelleContainer.cs are finished.

Reply:
I also did some testing. I put a print into the ChooseAction function to show number of mucocyst of the species that do have it. At the start of the game (first screenshot) all of them show 0 (some of them show 2 because their mucocysts split). After running the game for a longer time most species finally "understand" that they have mucocyst (second screenshot).
![image](https://github.com/user-attachments/assets/a1f6515f-d746-4919-aa26-b5a2065ae4f7)
![image](https://github.com/user-attachments/assets/551ca223-20ca-40d9-a08e-678ff6c5b210)

With my changes all of these cells know that they have mucocysts at the start (all of them show 1).
I also debugged the same way as you did. This function is called in the loading screen and all of them showed 0. When the game loaded only then did this value show a different number than 0.

Anyway in the `CalculateOrganelleLayoutStatistics` there is a line `container.OrganelleComponentsCached = false;` which makes it so that the `FetchLayoutOrganelleComponents` is called again... so maybe we should either remove it or move all organelles caching from `CalculateOrganelleLayoutStatistics` to `FetchLayoutOrganelleComponents` to reduce the number of recalculations

Reply:
I think you are seeing the species gain the mucocyst (so it at first doesn't have it but then it gets it and as old cell entities are despawned and freshly spawned ones replace them the mucocyst count goes up). I've debugged more today and I'm still convinced the existing code is correct. To verify, I started from master and added the following code to `MicrobeAIASystem.Update` and played the game for many minutes with the mucocyst:

```c#
        var species = (MicrobeSpecies)entity.Get<SpeciesMember>().Species;
        var mucocystCount = entity.Get<OrganelleContainer>().MucocystCount;
        var slime = SimulationParameters.Instance.GetOrganelleType("slimeJet");

        var speciesMucocystCount = species.Organelles.Count(o =>
            o.Definition.InternalName == slime.InternalName && o.Upgrades != null &&
            o.Upgrades.UnlockedFeatures.Contains("mucocyst"));

        if (mucocystCount < speciesMucocystCount)
        {
            GD.PrintErr($"Found entity with incorrect mucocyst count: {entity}");
            Debugger.Break();
        }
```

And I didn't hit the debugger break (or the print) so I'm very certain that mucocyst count calculation is correct. I did also verify that the species mucocyst count does go above 1 so that my testing code wasn't actually the bugged one. If you can show the problem appearing with that debugging code I wrote or if I have a bug in the debugging code then I'll believe the mucocyst count code has an issue. Though I think my testing code also has a potential "bug" where a species that gains a mucocyst while there are still old cells swimming around with the previous bodyplan. So I think it can report false positives, but it shouldn't be able to report a false negative.

Reply:
Okay, probably you're right about the spawning and despawning then, I will remove this mucocyst count update :)